libpod/container_internal_linux.go

```diff
@@ -744,7 +744,19 @@ func (c *Container) generateSpec(ctx context.Context) (*spec.Spec, error) {
 
 	// Warning: CDI may alter g.Config in place.
 	if len(c.config.CDIDevices) > 0 {
-		if err = cdi.UpdateOCISpecForDevices(g.Config, c.config.CDIDevices); err != nil {
+		registry := cdi.GetRegistry()
+		errs := registry.GetErrors()
+		if len(errs) > 0 {
+			logrus.Debugf("The following errors were triggered when creating the CDI Registry: %v", errs)
+		}
+		unresolved, err := registry.InjectDevices(g.Config, c.config.CDIDevices...)
+		if len(unresolved) > 0 {
+			logrus.Warnf("Could not resolve the following CDI devices: %v", unresolved)
+			if len(errs) > 0 {
+				logrus.Warnf("The following errors were triggered when creating the CDI Registry: %v", errs)
+			}
+		}
+		if err != nil {
```

Comment:
I think the err check should happen before the data check?

Reply:
If we return on `err != nil` we don't provide the warning / error output. With that said, I could include the output in the `if err != nil` block as there would (or at least should) not be a case where `unresolved` is non-empty and `err == nil`.

Reply:
OK. I have simplified things a bit. The error message will already show the unresolved devices, so there is no need to print them again. I am now also relying on the `Debugf` output to ensure that registry creation errors are available to the user instead of repeating them.